sgkit/tests/test_stats_utils.py

```diff
@@ -0,0 +1,162 @@
+
+from sgkit.stats.utils import (
+    assert_array_shape,
+    assert_block_shape,
+    assert_chunk_shape,
+    concat_2d,
+    r2_score,
+)
+
+
+def test_r2_score__batch_dims():
+    n, p, y = 20, 5, 3
+    np.random.seed(0)
+    X = np.random.normal(size=(n, p))
+    B = np.random.normal(size=(p, y))
+    Y = (X @ B).T
+    YP = Y + np.random.normal(size=(6, 8, y, n), scale=0.1)
+
+    # Test case with perfect predictions
+    np.testing.assert_allclose(r2_score(Y, Y), 1)
+
+    # Test case with near perfect predictions and extra
+    # loop dimensions
+    r2_actual = r2_score(YP, Y)
+    assert r2_actual.shape == YP.shape[:-1]
+    r2_expected = np.array(
+        [
+            r2_score(YP[i, j, k], Y[k])
+            for i in range(YP.shape[0])
+            for j in range(YP.shape[1])
+            for k in range(y)
+        ]
+    )
+    # This will ensure that aggregations occurred across
+    # the correct axis and that the loop dimensions can
+    # be matched with an explicit set of nested loops
+    np.testing.assert_allclose(r2_actual.ravel(), r2_expected)
+
+
+@pytest.mark.parametrize(  # type: ignore[misc]
+    "predicted,actual,expected_r2",  # type: ignore[no-untyped-def]
+    [
+        ([1, 1], [1, 2], -1.0),
+        ([1, 0], [1, 2], -7.0),
+        ([1, -1, 3], [1, 2, 3], -3.5),
+        ([0, -1, 2], [1, 2, 3], -4.5),
+        ([3, 2, 1], [1, 2, 3], -3.0),
+        ([0, 0, 0], [1, 2, 3], -6.0),
+        ([1.1, 2.1, 3.1], [1, 2, 3], 0.985),
+        ([1.1, 1.9, 3.0], [1, 2, 3], 0.99),
+        ([1, 2, 3], [1, 2, 3], 1.0),
+        ([1, 1, 1], [1, 1, 1], 1.0),
+        ([1, 1, 1], [1, 2, 3], -1.5),
+        ([1, 2, 3], [1, 1, 1], 0.0),
+    ],
+)
+def test_r2_score__sklearn_comparison(
+    predicted: List[Any], actual: List[Any], expected_r2: float
+):
+    yp, yt = np.array(predicted), np.array(actual)
+    assert r2_score(yp, yt) == expected_r2
+
+
+@given(st.integers(0, 25))
+@settings(max_examples=10)
+def test_concat_2d__values(n):
+    x, y = np.arange(n), np.arange(n * n).reshape(n, n)
+    z = np.copy(y)
+    ds = xr.Dataset(
+        dict(x=(("dim0"), x), y=(("dim0", "dim1"), y), z=(("dim1", "dim2"), z),)
+    )
+    actual = concat_2d(ds, dims=("dim0", "dim1"))
+    expected = np.concatenate([x.reshape(-1, 1), y], axis=1)
+    np.testing.assert_equal(actual, expected)
+
+
+def test_concat_2d__raise_on_gte_3d():
+    ds = xr.Dataset(dict(x=(("dim0", "dim1", "dim2"), np.empty((1, 1, 1)))))
+    with pytest.raises(ValueError, match="All variables must have <= 2 dimensions"):
+        concat_2d(ds, dims=("dim0", "dim1"))
+
+
+@st.composite
+def sample_dataset(draw):
+    # Draw a mixture of 1D or 2D arrays with identical length along
+    # first axis (second axis may be absent or have size 0)
+    n = draw(st.integers(0, 10))
+    n_1d = draw(st.integers(1, 5))
+    n_2d = draw(st.integers(0, 5))
+    n_cols = draw(st.tuples(*[st.integers(0, 5)] * n_2d))
+    shapes = n_1d * [(n,)] + [(n, nc) for nc in n_cols]
```

Comment:
There are some `numpy` specialised strategies in hypothesis - like `array_shapes`, the do not work here?